Document ZooKeeper credentials. Add missing examples.

stbof opened the pull request:
## Description
https://jira.mesosphere.com/browse/DOCS-2131

## Urgency
- [ ] Blocker <!-- Ping @sascala or @stbof for review -->
- [x ] High
- [ ] Medium

## Requirements
- Test all commands and procedures.
- Build content [locally](https://github.com/dcos/dcos-docs#test-local) and test for formatting/links.
- Add redirects to [dcos-website/redirect-files](https://github.com/dcos/dcos-website#managing-redirects).
- Change all affected versions (e.g. 1.7, 1.8, 1.9, and 1.10).
- See the [contribution guidelines](https://github.com/dcos/dcos-docs#contributing).

Commit:
Document ZooKeeper credentials. Add missing examples.

vespian commented:
@adamtheturtle Could you please also have a look? I can't add you as a reviewer :(

vespian commented on 1.10/installing/custom/configuration/examples.md:
```diff
@@ -130,8 +134,13 @@ https_proxy: https://<proxy_host>:<https_proxy_port>
 - '<blocked.address1.com>'
 - '<blocked.address2.com>'
+# Enterprise DC/OS Only
+zk_super_credentials: 'super:<long, random value>'
```
Should we explicitly say here that this is a string?

stbof commented:
@vespian  Joel left Mesosphere last summer.

vespian commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+### <a id="zk-agent"></a> zk_agent_credentials (Enterprise DC/OS Only)
+
+Credentials used to access the credentials of services that they will be running during the bootstrap process of the agent.
```
Can we please rephrase it somehow?

I would like to suggest: `Credentials used by bootstrapping code to access the credentials of the services that will be running on the agents.`

vespian commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+### <a id="zk-master"></a> zk_master_credentials (Enterprise DC/OS Only)
+
+Credentials used to access the credentials of services running on DC/OS Masters.
```
Same as below - can we please mention here that these creds will be used by the bootstraping code?

vespian reviewed (changes requested):
Left some comments.

vespian commented:
> @vespian Joel left Mesosphere last summer.
@stbof - @adamtheturtle is the GH handle of the member of the security team. He is very good at writing technical documents and I would like to get his opinion on this.

Commit:
Address feedback from @vespian.

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
 
 **Important:** You should also configure an HTTP proxy for [Docker](https://docs.docker.com/engine/admin/systemd/#/http-proxy).
+
+### <a id="zk-superuser"></a> zk_super_credentials (Enterprise DC/OS Only)
```
FTR I am just working on the assumption that these links are done in the right way.

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+### <a id="zk-superuser"></a> zk_super_credentials (Enterprise DC/OS Only)
+
+In `strict` and `permissive` clusters the information stored in ZooKeeper is protected by ACLs so that a malicious user can't connect to ZooKeeper quorum and directly modify service metadata. ACLs specify sets of IDs and permissions that are associated with those IDs. ZooKeeper supports pluggable authentication schemes and has a few built in schemes: `world`, `auth`, `digest`, `host`, `ip`. 
```
> In `strict` and `permissive` clusters

From the rest of the documentation, we say things like:

"On DC/OS clusters in strict and permissive mode...".

adamtheturtle replied:
The first time we use ACL in a section let's use it like we do in some of our other documentation:

> access control lists (ACLs)

adamtheturtle replied:
> protected by ACLs

ACLs don't protect anything.
Instead, perhaps "protected using ACLs"?

adamtheturtle replied:
> can't

Let's use "cannot" and ideally this would be in a style guide. For now it is my personal, internal style guide that I'm happy to discuss.

adamtheturtle replied:
> ZooKeeper quorum 

I believe that this should be "the ZooKeeper Quorum".

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+### <a id="zk-superuser"></a> zk_super_credentials (Enterprise DC/OS Only)
+
+In `strict` and `permissive` clusters the information stored in ZooKeeper is protected by ACLs so that a malicious user can't connect to ZooKeeper quorum and directly modify service metadata. ACLs specify sets of IDs and permissions that are associated with those IDs. ZooKeeper supports pluggable authentication schemes and has a few built in schemes: `world`, `auth`, `digest`, `host`, `ip`. 
```
> ACLs specify sets of IDs and permissions that are associated with those IDs

Rather than "IDs and permissions" I believe that these are resource IDs (RIDs) and actions".

adamtheturtle replied:
> `world`, `auth`, `digest`, `host`, `ip`. 

Let's put "and" before the last item in this list.

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+In `strict` and `permissive` clusters the information stored in ZooKeeper is protected by ACLs so that a malicious user can't connect to ZooKeeper quorum and directly modify service metadata. ACLs specify sets of IDs and permissions that are associated with those IDs. ZooKeeper supports pluggable authentication schemes and has a few built in schemes: `world`, `auth`, `digest`, `host`, `ip`. 
+
+DC/OS ZooKeeper credentials `zk_super_credentials`, `zk_master_credentials`, and `zk_agent_credentials` use `digest` authentication, which require a `<uid>:<password>` string which is then used as an ID while checking if a client can access a particular resource.
```
> use `digest` authentication, which require a [...] string

This isn't grammatically correct. If the meaning is that the authentication itself requires a string, change "require" to "requires".

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+DC/OS ZooKeeper credentials `zk_super_credentials`, `zk_master_credentials`, and `zk_agent_credentials` use `digest` authentication, which require a `<uid>:<password>` string which is then used as an ID while checking if a client can access a particular resource.
+
+`zk_super_credentials` enables access to ZooKeeper's equivalent of `root/superuser` account, which has access to all resources regardless of existing ACLs. This credential allows an operator to access all the metadata stored in the ZooKeeper quorum and is used by the DC/OS bootstrap script while initializing the cluster. Default: `'super:secret'`
```
> equivalent of `root/superuser` account

either "of a" or "of the".

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
````diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+To harden clusters, Mesosphere recommends that you change the defaults of all credentials to long, complex values. Once set, you can verify the settings using `/opt/mesosphere/active/exhibitor/usr/zookeeper/bin/zkCli.sh` available on DC/OS Master nodes. By default, zkCli does not authenticate, so the nodes in `/dcos` tree will not be accessible. After invoking `addauth digest <zk_super_credential ...
+
+```
````
Can this have some highlighting? Perhaps:

````
```console
...
```
````

?

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+`zk_super_credentials` enables access to ZooKeeper's equivalent of `root/superuser` account, which has access to all resources regardless of existing ACLs. This credential allows an operator to access all the metadata stored in the ZooKeeper quorum and is used by the DC/OS bootstrap script while initializing the cluster. Default: `'super:secret'`
+
+To harden clusters, Mesosphere recommends that you change the defaults of all credentials to long, complex values. Once set, you can verify the settings using `/opt/mesosphere/active/exhibitor/usr/zookeeper/bin/zkCli.sh` available on DC/OS Master nodes. By default, zkCli does not authenticate, so the nodes in `/dcos` tree will not be accessible. After invoking `addauth digest <zk_super_credential ...
```
> DC/OS Master nodes

The rest of our documentation seems to use lowercase "m" for "master". e.g. "Admin Router runs on DC/OS master nodes".

adamtheturtle replied:
Change `zkCli` to `` `zkCli` ``.

adamtheturtle replied:
> the nodes in `/dcos` tree

This is not grammatically correct. I guess change "in" to "in the"?

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,29 @@ Indicates whether to enable the DC/OS proxy.
+
+### <a id="zk-master"></a> zk_master_credentials (Enterprise DC/OS Only)
+
+Credentials used by bootstrapping code to access the credentials of the services that will be running on the masters.
```
❓ Is it relevant that it is code? We don't say "code" elsewhere that I can find to refer to our own processes. Perhaps "used by the bootstrap process", but I'm not sure.

> the masters

Let's be clear and say "DC/OS master nodes" or whatever the right distinction is.

adamtheturtle reviewed (changes requested):
Thank you @stbof . Please resubmit after the requested changes have been made, and feel free to ask the security team for help on specifics 😄

Commit:
Address feedback from @adamtheturtle.

stbof commented:
@adamtheturtle  I believe I have addressed all your feedback.  Our current web site doesn't support highlighting of console output, but our new doc website will.

Commit:
Add zk cred to 1.9 and fix jump links.

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,31 @@ Indicates whether to enable the DC/OS proxy.
+
+DC/OS ZooKeeper credentials `zk_super_credentials`, `zk_master_credentials`, and `zk_agent_credentials` use `digest` authentication, which requires a `<uid>:<password>` string which is then used as an ID while checking if a client can access a particular resource.
+
+`zk_super_credentials` enables access to ZooKeeper's equivalent of the `root/superuser` account, which has access to all resources regardless of existing ACLs. This credential allows an operator to access all the metadata stored in the ZooKeeper quorum and is used by the DC/OS bootstrap script while initializing the cluster. Default: `'super:secret'`.
```
I believe "quorum" should be "Quorum".
I'm sure that it should be consistent within this document.

adamtheturtle replied:
This implies that there is an account called "root/superuser" at least in some system. Instead, some systems use "root", some use "superuser" and this account is equivalent to both.

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
````diff
@@ -501,3 +501,31 @@ Indicates whether to enable the DC/OS proxy.
+
+To harden clusters, Mesosphere recommends that you change the defaults of all credentials to long, complex values. Once set, you can verify the settings using `/opt/mesosphere/active/exhibitor/usr/zookeeper/bin/zkCli.sh` available on DC/OS master nodes. By default, `zkCli` does not authenticate, so the nodes in the `/dcos` tree will not be accessible. After invoking `addauth digest <zk_super_cred ...
+
+```powershell
````
Is this definitely powershell?

adamtheturtle replied:
I'm no PowerShell expert, but I think that it isn't, as this includes `ls` when PowerShell uses `dir`. Instead, this is `console`.

The commands alone are either `bash` or `sh`, but as this shows the prompt, it is `console` which is generic.

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,31 @@ Indicates whether to enable the DC/OS proxy.
+<a id="zk-agent"></a> 
+### zk_agent_credentials (Enterprise DC/OS Only)
+
+Credentials used by bootstrapping process to access the credentials of the services that will be running on the DC/OS agent nodes.
```
Missing article before "bootstrapping".

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,31 @@ Indicates whether to enable the DC/OS proxy.
+<a id="zk-master"></a> 
+### zk_master_credentials (Enterprise DC/OS Only)
+
+Credentials used by bootstrapping process to access the credentials of the services that will be running on the DC/OS master nodes.
```
Missing article before "bootstrapping" - I guess "the bootstrapping process" but @vespian might know the used term better than I do.

vespian replied:
`the bootstrapping processes` please :)

adamtheturtle commented on 1.10/installing/custom/configuration/configuration-parameters.md:
```diff
@@ -501,3 +501,31 @@ Indicates whether to enable the DC/OS proxy.
 **Important:** You should also configure an HTTP proxy for [Docker](https://docs.docker.com/engine/admin/systemd/#/http-proxy).
+
+<a id="zk-superuser"></a> 
+### zk_super_credentials (Enterprise DC/OS Only)
```
This is now "DC/OS Enterprise" and not "Enterprise DC/OS" I believe.

stbof replied:
@adamtheturtle Yes, but we are not doing this wholesale renaming until we deploy the new merged DC/OS Enterprise and DC/OS open source doc sites.

adamtheturtle reviewed (changes requested):
Thanks @stbof - I've spotted all that I will today 😄

Commit:
Incorporate feedback from @vespian and @adamtheturtle.

Commit:
Merge branch 'master' into sb/DOCS-2131
